bftclient/src/bft_client.cpp

```diff
@@ -153,45 +184,82 @@ Reply Client::send(const MatchConfig& match_config,
       communication_->send(dests, std::move(msg));
     }
 
-    if (auto reply = wait()) {
+    wait(replies);
+    if (replies.size() == 1) {
       expected_commit_time_ms_.add(
           std::chrono::duration_cast<std::chrono::milliseconds>(std::chrono::steady_clock::now() - start).count());
-      return reply.value();
+      replyCertificates_.clear();
+      return replies.front();
     }
     metrics_.retransmissions.Get().Inc();
   }
 
   expected_commit_time_ms_.add(request_config.timeout.count());
-  outstanding_request_ = std::nullopt;
+  replyCertificates_.clear();
   throw TimeoutException(request_config.sequence_number, request_config.correlation_id);
 }
 
-std::optional<Reply> Client::wait() {
+std::deque<Reply> Client::sendBatch(std::deque<ClientRequest>& clientRequests, const std::string& cid) {
+  std::deque<Reply> replies;
+  std::chrono::milliseconds max_time_to_wait = 0s;
+  Msg client_msg;
+  MatchConfig match_config;
+  uint32_t max_reply_size = 0;
+  for (auto& req : clientRequests) {
+    match_config = writeConfigToMatchConfig(req.config);
+    replyCertificates_.insert(std::make_pair(req.config.request.sequence_number, Matcher(match_config)));
+    pendingRequests_.push_back(createClientMsg(req.config.request, std::move(req.request), false, config_.id.val));
+    if (req.config.request.timeout > max_time_to_wait) max_time_to_wait = req.config.request.timeout;
+    if (req.config.request.max_reply_size > max_reply_size) max_reply_size = req.config.request.max_reply_size;
+  }
+  receiver_.activate(max_reply_size);
+  uint32_t batchBufSize = 0;
+  for (auto const& msg : pendingRequests_) batchBufSize += msg.size();
+  auto start = std::chrono::steady_clock::now();
+  auto end = start + max_time_to_wait;
+  auto batch_msg = createClientBatchMsg(pendingRequests_, batchBufSize, cid, config_.id.val);
+  while (std::chrono::steady_clock::now() < end && replies.size() != pendingRequests_.size()) {
+    bft::client::Msg msg(batch_msg);  // create copy here due to the loop
+    if (primary_) {
+      communication_->send(primary_.value().val, std::move(msg));
+    } else {
+      std::set<bft::communication::NodeNum> dests;
+      for (const auto& d : match_config.quorum.destinations) {
+        dests.emplace(d.val);
+      }
+      communication_->send(dests, std::move(msg));
+    }
+
+    wait(replies);
+    metrics_.retransmissions.Get().Inc();
+  }
+  replyCertificates_.clear();
+  if (replies.size() == pendingRequests_.size()) return replies;
+
+  expected_commit_time_ms_.add(max_time_to_wait.count());
+  replyCertificates_.clear();
```

Comment:
This call is a duplicate of the call on line 236.

Reply:
Done.